base/abstractarraymath.jl

```diff
@@ -487,11 +525,21 @@ julia> collect(eachslice(M, dims=2))
  [3, 6, 9]
 ```
 """
-@inline function eachslice(A::AbstractArray; dims)
-    length(dims) == 1 || throw(ArgumentError("only single dimensions are supported"))
-    dim = first(dims)
-    dim <= ndims(A) || throw(DimensionMismatch("A doesn't have $dim dimensions"))
-    inds_before = ntuple(d->(:), dim-1)
-    inds_after = ntuple(d->(:), ndims(A)-dim)
-    return (view(A, inds_before..., i, inds_after...) for i in axes(A, dim))
+@inline eachslice(A;dims) = _eachslice(A,dims)
+
+@inline _eachslice(A,dim::Integer) = _eachslice(A,(dim,))
+@inline function _eachslice(A::AbstractArray{N}, dims::NTuple{M,Integer}) where {N,M}
+    for (i,dim) in enumerate(dims)
+        1 <= dim <= N || throw(DimensionMismatch("A doesn't have dimension $dim"))
+        dim in dims[1:i-1] && throw(DimensionMismatch("$dims are not unique"))
+    end
+    iter = CartesianIndices(map(dim -> axes(A,dim), dims))
+    lookup = ntuple(dim -> findfirst(isequal(dim), dims), N)
+
+    # determine element type
+    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : eltype(a), axes(A), lookup)...}
```

Comment:
We need to explicitly communicate to SubArray that these are going to be slices over the entire dimension (that helps make things linear more frequently).

```suggestion
    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : Base.Slice{eltype(a)}, axes(A), lookup)...}
```